repository/repository.go

```diff
@@ -470,3 +475,159 @@ To find more information on specific steps, please see documentation [here](http
 - [ ] QA: Final validation of above PR and tracked through the linked ticket
 - [ ] PJM: Close the milestone in GitHub.
 `
+
+// GetUpstreamRemote will return the remote name for a given configured remote URL
+func GetUpstreamRemote(r *git.Repository, rURL string) (string, error) {
+	remotes, err := r.Remotes()
+	if err != nil {
+		return "", err
+	}
+
+	for _, remote := range remotes {
+		if remote.Config().URLs[0] == rURL {
+			return remote.Config().Name, nil
+		}
+	}
+
+	return "", errors.New("upstream remote not found")
+}
+
+// DiffLocalToRemote will get the commits from the local branch and from the target remote branch,
+// will return the commits that are present in the local branch but not in the remote branch.
+func DiffLocalToRemote(r *git.Repository, remote, rb string) error {
+	var (
+		refSpec   string
+		localRef  *plumbing.Reference
+		remoteRef *plumbing.Reference
+		err       error
+	)
+
+	refSpec = "refs/heads/" + rb + ":refs/remotes/" + remote + "/" + rb
+
+	fetchOpts := &git.FetchOptions{
+		RemoteName: remote,
+		RefSpecs:   []config.RefSpec{config.RefSpec(refSpec)},
+		Force:      true,
+	}
+
+	// Fetch the remote branch
+	if err = r.Fetch(fetchOpts); err != nil && err != git.NoErrAlreadyUpToDate {
+		return err
+	}
+
+	// Get the local and remote branch references
+	localRef, err = r.Head()
+	if err != nil {
+		return err
+	}
+
+	remoteRef, err = r.Reference(plumbing.NewRemoteReferenceName(remote, rb), true)
+	if err != nil {
+		return err
+	}
+
+	// Find the commits that are in the local branch but not in the remote branch
+	uniqueCommits, err := findUniqueCommits(r, localRef, remoteRef)
+	if err != nil {
+		return err
+	}
+
+	// Print the commit message and hash for each unique commit
+	for _, commit := range uniqueCommits {
+		fmt.Printf("Commit: %s\nMessage: %s\n\n", commit.Hash, commit.Message)
+	}
+
+	return nil
+}
+
+// PushRemoteBranch will push the local branch to the remote repository
+func PushRemoteBranch(r *git.Repository, remote, u, t string, debug bool) error {
+	var (
+		branchRef string
+		h         *plumbing.Reference
+	)
+
+	h, err := r.Head()
+	if err != nil {
+		return err
+	}
+
+	branchRef = h.Name().String()
+
+	auth := &http.BasicAuth{
+		Username: u,
+		Password: t,
+	}
+
+	opts := &git.PushOptions{
+		RemoteName: remote,
+		Force:      true,
+		RefSpecs:   []config.RefSpec{config.RefSpec(branchRef + ":" + branchRef)},
+		Auth:       auth,
+	}
+
+	if debug {
+		opts.Progress = os.Stdout
+	}
+
+	return r.Push(opts)
+}
```

Comment:
Does using basic auth with 2FA on works?

Reply:
I have 2FA configured in my account for some actions. 

Pushing does not require me to authenticate with 2FA. 
So I don't know.
I will see with you in private, let's test this. 

However there is something that this reminded me:
```
// PushRemoteBranch will push the local branch to the remote repository
func PushRemoteBranch(r *git.Repository, remote, u, t string, debug bool) error {
	var (
		branchRef string
		h         *plumbing.Reference
	)

	h, err := r.Head()
	if err != nil {
		return err
	}

	branchRef = h.Name().String()

	auth := &http.BasicAuth{
		Username: u,
		Password: t,
	}

	opts := &git.PushOptions{
		RemoteName: remote,
		Force:      true,
		RefSpecs:   []config.RefSpec{config.RefSpec(branchRef + ":" + branchRef)},
		Auth:       auth,
	}

	if debug {
		opts.Progress = os.Stdout
	}

	return r.Push(opts)
}
```
The push is only authenticating with `token + user`, what is the default strategy we use on this repo? 
ssh + token & user? 

@tashima42

Reply:
I would guess that 2FA isn't a problem since it is a token with a set expiration date, other places use the ssh key, but honestly, I don't think one is necessarily better than the other.
https://github.com/rancher/ecm-distro-tools/blob/5f7ffe5b63a06fc858a391e8ccc6c97a934523ab/release/k3s/k3s.go#L269-L287